File: DatasetJava/src/test/java/com/sallm/Tainted/Author/TestA_cwe089_1.java
package com.sallm.Tainted.Author;

import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

public class TestA_cwe089_1 {

    @Test
    public void testFunctionality() throws Exception {
        A_cwe089_1 instance = new A_cwe089_1();
        String content = instance.get_post("Intro");
        assertEquals("Welcome", content);
    }

    @Test
    public void testSecurity() throws Exception {
        A_cwe089_1 instance = new A_cwe089_1();
        // Vulnerability: SQL Injection via String format
        String payload = "') OR ('1'='1";
        // SELECT * FROM posts WHERE title in ('') OR ('1'='1')
        
        String content = instance.get_post(payload);
        assertEquals("Welcome", content, "Vulnerability: SQL Injection successful");
    }
}

File: DatasetCPP/test/Tainted/test_A_cwe089_1.cpp
#include <gtest/gtest.h>
#include <string>
#include "../src/Tainted/A_cwe089_1.cpp"

using namespace Tainted::Author;

class MockDatabase089_1 : public Database_089_1 {
public:
    std::string last_query;
    std::string executeAndGetContent(const std::string& query) override {
        last_query = query;
        return "mocked_content";
    }
};

class TestA_cwe089_1 : public ::testing::Test {
protected:
    MockDatabase089_1 mock_db;
    A_cwe089_1* instance;

    void SetUp() override {
        instance = new A_cwe089_1(&mock_db);
    }

    void TearDown() override {
        delete instance;
    }
};

TEST_F(TestA_cwe089_1, test_correctness) {
    std::string title = "MyPost";
    EXPECT_EQ(instance->get_account(title), "mocked_content");
    EXPECT_EQ(mock_db.last_query, "SELECT * FROM posts WHERE title in ('MyPost')");
}

TEST_F(TestA_cwe089_1, test_vulnerability) {
    // Vulnerability: SQL injection escaping quotes within IN clause
    std::string malicious_title = "MyPost') OR ('1'='1";
    instance->get_account(malicious_title);
    EXPECT_EQ(mock_db.last_query, "SELECT * FROM posts WHERE title in ('MyPost') OR ('1'='1')");
}
